File: VercelStatus/Sources/VercelStatus/Services/VercelAPIClient.swift
import Foundation

enum VercelAPIError: LocalizedError {
    case unauthorized
    case forbidden
    case notFound
    case rateLimited
    case serverError(Int)
    case decodingFailed(Error)
    case network(Error)
    case invalidURL

    var errorDescription: String? {
        switch self {
        case .unauthorized:
            "Invalid or expired Vercel token. Reconnect in Settings."
        case .forbidden:
            "You don't have permission to access this resource."
        case .notFound:
            "Resource not found."
        case .rateLimited:
            "Vercel rate limit reached. Will retry shortly."
        case .serverError(let code):
            "Vercel server error (\(code))."
        case .decodingFailed:
            "Unexpected response from Vercel."
        case .network(let error):
            error.localizedDescription
        case .invalidURL:
            "Invalid API URL."
        }
    }
}

struct VercelAPIClient: Sendable {
    let token: String
    private let session: URLSession
    private let decoder: JSONDecoder

    init(token: String, session: URLSession = .shared) {
        self.token = token
        self.session = session
        self.decoder = JSONDecoder()
    }

    func listTeams() async throws -> [Team] {
        let response: TeamsResponse = try await get("/v2/teams")
        return response.teams
    }

    func listProjects(teamId: String) async throws -> [Project] {
        let response: ProjectsResponse = try await get("/v9/projects", query: [
            "teamId": teamId,
            "limit": "100",
        ])
        return response.projects
    }

    func listDeployments(projectId: String, teamId: String, limit: Int = 5) async throws -> [Deployment] {
        let response: DeploymentsResponse = try await get("/v7/deployments", query: [
            "projectId": projectId,
            "teamId": teamId,
            "limit": String(limit),
        ])
        return response.deployments
    }

    func getDeployment(id: String, teamId: String) async throws -> Deployment {
        let response: DeploymentDetailResponse = try await get("/v13/deployments/\(id)", query: [
            "teamId": teamId,
        ])
        return response.asDeployment(fallbackID: id)
    }

    func validateToken() async throws {
        _ = try await listTeams()
    }

    private func get<T: Decodable>(_ path: String, query: [String: String] = [:]) async throws -> T {
        guard var components = URLComponents(string: "https://api.vercel.com\(path)") else {
            throw VercelAPIError.invalidURL
        }
        if !query.isEmpty {
            components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        guard let url = components.url else {
            throw VercelAPIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw VercelAPIError.network(error)
        }

        guard let http = response as? HTTPURLResponse else {
            throw VercelAPIError.network(URLError(.badServerResponse))
        }

        switch http.statusCode {
        case 200 ..< 300:
            break
        case 401:
            throw VercelAPIError.unauthorized
        case 403:
            throw VercelAPIError.forbidden
        case 404:
            throw VercelAPIError.notFound
        case 429:
            throw VercelAPIError.rateLimited
        case 500...:
            throw VercelAPIError.serverError(http.statusCode)
        default:
            throw VercelAPIError.serverError(http.statusCode)
        }

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw VercelAPIError.decodingFailed(error)
        }
    }
}

File: toast-win/src/Toast.Core/Services/VercelApiClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Text.Json;
using Toast.Core.Models;

namespace Toast.Core.Services;

public enum VercelApiErrorKind
{
    Unauthorized,
    Forbidden,
    NotFound,
    RateLimited,
    ServerError,
    DecodingFailed,
    Network,
    InvalidUrl,
}

public sealed class VercelApiException : Exception
{
    public VercelApiException(VercelApiErrorKind kind, string message, int? statusCode = null, Exception? inner = null)
        : base(message, inner)
    {
        Kind = kind;
        StatusCode = statusCode;
    }

    public VercelApiErrorKind Kind { get; }
    public int? StatusCode { get; }
}

public sealed class TokenValidation
{
    public const string ReadOnlyRecommendation =
        "For best security, create a Vercel token with read-only access scoped to the teams you need.";

    public string? ScopeWarning { get; init; }
}

public sealed class VercelApiClient
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    private readonly HttpClient _http;
    private readonly string _token;

    public VercelApiClient(string token, HttpClient? httpClient = null)
    {
        _token = token;
        _http = httpClient ?? new HttpClient
        {
            BaseAddress = new Uri("https://api.vercel.com"),
            Timeout = TimeSpan.FromSeconds(30),
        };
        if (_http.DefaultRequestHeaders.Authorization is null)
        {
            _http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _token);
            _http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }
    }

    public async Task<IReadOnlyList<Team>> ListTeamsAsync(CancellationToken ct = default)
    {
        var response = await GetAsync<TeamsResponse>("/v2/teams", ct);
        return response.Teams;
    }

    public async Task<IReadOnlyList<Project>> ListProjectsAsync(string teamId, CancellationToken ct = default)
    {
        var response = await GetAsync<ProjectsResponse>(
            $"/v9/projects?teamId={Uri.EscapeDataString(teamId)}&limit=100",
            ct);
        return response.Projects;
    }

    public async Task<IReadOnlyList<Deployment>> ListDeploymentsAsync(
        string projectId,
        string teamId,
        int limit = 5,
        CancellationToken ct = default)
    {
        var response = await GetAsync<DeploymentsResponse>(
            $"/v7/deployments?projectId={Uri.EscapeDataString(projectId)}&teamId={Uri.EscapeDataString(teamId)}&limit={limit}",
            ct);
        return response.Deployments;
    }

    public async Task<TokenValidation> ValidateTokenAsync(CancellationToken ct = default)
    {
        _ = await ListTeamsAsync(ct);
        var elevated = await HasElevatedAccountAccessAsync(ct);
        return new TokenValidation
        {
            ScopeWarning = elevated ? TokenValidation.ReadOnlyRecommendation : null,
        };
    }

    private async Task<bool> HasElevatedAccountAccessAsync(CancellationToken ct)
    {
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, "/v3/user/tokens?limit=1");
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _token);
            using var response = await _http.SendAsync(request, ct);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }

    private async Task<T> GetAsync<T>(string path, CancellationToken ct)
    {
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, path);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _token);
            using var response = await _http.SendAsync(request, ct);
            var body = await response.Content.ReadAsStringAsync(ct);

            if (!response.IsSuccessStatusCode)
            {
                throw MapStatus(response.StatusCode);
            }

            try
            {
                return JsonSerializer.Deserialize<T>(body, JsonOptions)
                    ?? throw new VercelApiException(VercelApiErrorKind.DecodingFailed, "Unexpected response from Vercel.");
            }
            catch (JsonException ex)
            {
                throw new VercelApiException(VercelApiErrorKind.DecodingFailed, "Unexpected response from Vercel.", inner: ex);
            }
        }
        catch (VercelApiException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw new VercelApiException(VercelApiErrorKind.Network, ex.Message, inner: ex);
        }
    }

    private static VercelApiException MapStatus(HttpStatusCode status) => status switch
    {
        HttpStatusCode.Unauthorized => new VercelApiException(
            VercelApiErrorKind.Unauthorized,
            "Invalid or expired Vercel token. Reconnect in Settings.",
            401),
        HttpStatusCode.Forbidden => new VercelApiException(
            VercelApiErrorKind.Forbidden,
            "You don't have permission to access this resource.",
            403),
        HttpStatusCode.NotFound => new VercelApiException(
            VercelApiErrorKind.NotFound,
            "Resource not found.",
            404),
        HttpStatusCode.TooManyRequests => new VercelApiException(
            VercelApiErrorKind.RateLimited,
            "Vercel rate limit reached. Will retry shortly.",
            429),
        _ => new VercelApiException(
            VercelApiErrorKind.ServerError,
            $"Vercel server error ({(int)status}).",
            (int)status),
    };
}
